doc/rfc/781-lua-system-wide-profiler.md

```diff
@@ -0,0 +1,395 @@
+# lua: system-wide profiler
+
+* **Status**: Done
+* **Start date**: 02-07-2021
+* **Authors**: Mikhail Shishatskiy @Shishqa m.shishatskiy@tarantool.org, Maxim Kokryashkin @fckxorg m.kokryashkin@tarantool.org
+* **Issues**: [#781](https://github.com/tarantool/tarantool/issues/781), [#4001](https://github.com/tarantool/tarantool/issues/4001)
+
+## Summary
+The document describes the platform profiler for LuaJIT. It is needed to obtain a complete view of platform performance. The existing LuaJIT profiler is only able to give you information about virtual machine states and guest stacks. Hence, the document proposes to provide a new sampling profiler, so it will be able to gather stack traces from both C and Lua.
+
+## Background and motivation
+
+Currently, the available options for profiling LuaJIT are not fine enough to provide an understanding of performance. For example, perf is only able to show the host stack, so all the Lua calls are seen as a single pcall. Oppositely, the jit.p module provided with LuaJIT is not able to give any information about the host stack.
+
+To get a detailed perspective on the platform's performance, a more advanced profiler is needed. The desired profiler must be able to capture both guest and host stacks simultaneously, along with virtual machine states.
+
+Assume we have the following chunks of code:
+```c
+1  // file c_payload.c
+2
+3  // fibonnacci number
+4  double c_payload(double n) {
+5    if (n <= 1) {
+6      return n;
+7    }
+8    return c_payload(n - 1) + c_payload(n - 2);
+9  }
+```
+
+```lua
+ 1  -- file lua_payload.lua
+ 2
+ 3  local jit = require("jit")
+ 4  local ffi = require("ffi")
+ 5  local os = require("os")
+ 6
+ 7  ------------------------------------------------
+ 8
```

Comment:
I noticed, that we have `@../lua_lib.lua:8` on flamegraph, but this it should be `:10` shouldn't it?

Reply:
I have no strict opinion for this: there would be nice to use the actual flamegraph here, but I believe we can ignore this (or fix on demand if someone is confused by it). If everybody is OK, I'll simply merge this PR as is.